### PR TITLE
Launch chrome as a part of Lighthouse CLI

### PR DESCRIPTION
**Pull request by boopathi**
#349 

+ ~~Rebased the old branch~~
+ ~~Works only for OSX now~~

--- 

+ flag `--select-chrome` to choose from chrome installation when multiple are found
+ ~~flag `--no-launch-chrome` to disable chrome launch~~
+ flag `--skip-autolaunch` to disable chrome launch

### CODE REVIEW

File: lighthouse-cli/chrome-launcher.js [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+    this.head = defaults(opts.head, true);
+    this.osxExecPath = defaults(opts.osxExecPath, '/Contents/MacOS/Google Chrome Canary');
+
+    switch (process.platform) {
+      case 'darwin':
+        this.TMP_PROFILE_DIR = execSync('mktemp -d -t lighthouse').toString().trim();
+        this.run = this.spawnOsx.bind(this);
+        break;
+      case 'linux':
+        this.TMP_PROFILE_DIR = execSync('mktemp -d -t lighthouse.XXXXXXX').toString().trim();
```
Comment (paulirish) [inline lighthouse-cli/chrome-launcher.js]: both mac & linux can use this one. the XXXX's don't hurt.

File: lighthouse-cli/chrome-launcher.js [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+        fs.writeSync(this.pidFile, chrome.pid.toString());
+
+        console.log('chrome running with pid =', chrome.pid);
+        return this.poll(0).then(_ => chrome.pid);
+      });
+  }
+
+  cleanup(client) {
+    if (client) {
+      client.removeAllListeners('connect');
```
Comment (paulirish) [inline lighthouse-cli/chrome-launcher.js]: call without args to nuke them all?

**Comment by boopathi** [comment]
Added some linux and windows stuff and I tried it out on OSX and Windows 10.

**Comment by paulirish** [comment]
We're pretty conservative with runtime dependencies. 

Inquirer is pretty heavyweight: http://npm.anvaka.com/#/view/2d/inquirer 
do we need it?

also, same concern with through2.

File: lighthouse-cli/index.js [hunk 3]
```diff
@@ -60,7 +61,8 @@ const cli = yargs
     'save-artifacts': 'Save all gathered artifacts to disk',
     'list-all-audits': 'Prints a list of all available audits and exits',
     'list-trace-categories': 'Prints a list of all required trace categories and exits',
-    'config-path': 'The path to the config JSON.'
+    'config-path': 'The path to the config JSON.',
+    'select-chrome': 'Choose chrome location to use when multiple installations are found',
```
Comment (paulirish) [inline lighthouse-cli/index.js]: Interesting. How does this flag work?

**Comment by paulirish** [comment]
Overall this patch is very well defined and I'm excited to get it landed. :) A few things we'll want to work through first, though.

File: lighthouse-cli/osx-chrome.js [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+      .map(inst => {
+        for (let [regex, priority] of priorities) {
+          if (regex.test(inst)) {
+            return [inst, priority];
+          }
+        }
+        return [inst, defaultPriority];
+      })
+      // sort based on priorities
+      .sort((a, b) => {
```
Comment (paulirish) [inline lighthouse-cli/osx-chrome.js]: I believe you can simplify this to 

```js
.sort((a, b) => b[1] - a[1])
```

**Comment by paulirish** [comment]
Can we add a flag for `--dont-launch-chrome` or something? 

This would allow someone to use lighthouse against a target that already has 9222 open, such as when remote debugging a real device: https://github.com/GoogleChrome/lighthouse/issues/219

File: lighthouse-cli/chrome-launcher.js [hunk 5]
```diff
@@ -0,0 +1,187 @@
+'use strict';
```
Comment (brendankenny) [inline lighthouse-cli/chrome-launcher.js]: need license

File: lighthouse-cli/chrome-launcher.js [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+          process.env.TMP ||
+          (process.env.SystemRoot || process.env.windir) + '\\temp';
+        const randomNumber = Math.floor(Math.random() * 9e7 + 1e7);
+
+        this.TMP_PROFILE_DIR = path.join(tmpdir, 'lighthouse' + randomNumber);
+        mkdirp.sync(this.TMP_PROFILE_DIR);
+        break;
+
+      default:
+        throw platformError;
```
Comment (brendankenny) [inline lighthouse-cli/chrome-launcher.js]: maybe just check in the constructor so you don't have to check for an unsupported platform in more than one place?

File: lighthouse-cli/chrome-launcher.js [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+    }
+
+    this.outFile = fs.openSync(`${this.TMP_PROFILE_DIR}/chrome-out.log`, 'a');
```
Comment (brendankenny) [inline lighthouse-cli/chrome-launcher.js]: are these used or could you just `'ignore'`?

File: lighthouse-cli/chrome-launcher.js [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+
+function defaults(val, def) {
+  return typeof val === 'undefined' ? def : val;
+}
+
+module.exports = class Launcher {
+  constructor(opts) {
+    opts = opts || {};
+    // choose the first one (default)
+    this.head = defaults(opts.head, true);
```
Comment (brendankenny) [inline lighthouse-cli/chrome-launcher.js]: maybe a more descriptive name for this than `head`? It's not immediately clear what that would mean.

File: lighthouse-cli/chrome-launcher.js (hunk 6 above)
Comment (boopathi) [inline lighthouse-cli/chrome-launcher.js]: this was to have a default case in switch for the linter. It was enough to check only in `prepare` function.

File: lighthouse-cli/chrome-launcher.js (hunk 7 above)
Comment (boopathi) [inline lighthouse-cli/chrome-launcher.js]: chrome stdout and stderr are logged to these files.

File: lighthouse-cli/chrome-launcher.js (hunk 8 above)
Comment (boopathi) [inline lighthouse-cli/chrome-launcher.js]: yeah. Should do that .. any suggestions ?

Comment (brendankenny) [inline lighthouse-cli/chrome-launcher.js]: is the idea it picks the first instance found? How will you know which one lighthouse will find first?

File: lighthouse-cli/index.js (hunk 3 above)
Comment (boopathi) [inline lighthouse-cli/index.js]: Maybe, with that we could remove inquirer.

File: lighthouse-cli/chrome-launcher.js (hunk 6 above)
Comment (brendankenny) [inline lighthouse-cli/chrome-launcher.js]: but it's used in `prepare()` and `run()`. Obviously the platform won't change between calls to each :)

File: lighthouse-cli/index.js (hunk 3 above)
Comment (boopathi) [inline lighthouse-cli/index.js]: letting user choose from a list seemed more easy than specifying a path.

File: lighthouse-cli/chrome-launcher.js (hunk 7 above)
Comment (brendankenny) [inline lighthouse-cli/chrome-launcher.js]: right, but then we delete the temp directory, so the user won't have access. Is there any reason not to ignore?

File: lighthouse-cli/chrome-launcher.js (hunk 8 above)
Comment (boopathi) [inline lighthouse-cli/chrome-launcher.js]: I have a priority list that will sort the instances based on priorities assigned and then from the sorted list, it chooses the first one.

File: lighthouse-cli/chrome-launcher.js (hunk 6 above)
Comment (boopathi) [inline lighthouse-cli/chrome-launcher.js]: yes. It's only there for the linter to pass.

File: lighthouse-cli/chrome-launcher.js (hunk 7 above)
Comment (boopathi) [inline lighthouse-cli/chrome-launcher.js]: if chrome did not start properly, there won't be anywhere to check, and lighthouse logs won't report why chrome didn't start. Right now it is in a separate terminal tab.

File: lighthouse-cli/chrome-launcher.js (hunk 6 above)
Comment (brendankenny) [inline lighthouse-cli/chrome-launcher.js]: yeah, it may make sense not to start preparing until `run()` is actually called anyways.

File: lighthouse-cli/chrome-launcher.js (hunk 7 above)
Comment (brendankenny) [inline lighthouse-cli/chrome-launcher.js]: right, but the chrome logs are still deleted in case of error, aren't they?

Comment (brendankenny) [inline lighthouse-cli/chrome-launcher.js]: (logging why Chrome is failing is good, I'm just saying it doesn't seem like you'd be able to get the logs if it does fail)

File: lighthouse-cli/chrome-launcher.js [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+
+    this.pollInterval = 500;
+
+    this.chromeInstances = [];
```
Comment (brendankenny) [inline lighthouse-cli/chrome-launcher.js]: won't there only ever be one chrome instance?

File: lighthouse-cli/osx-chrome.js [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+        }
+        return 0;
+      })
+      // remove priority flag
+      .map(pair => pair[0]);
+  }
+
+  getChrome() {
+    return new Promise((resolve, reject) => {
+      const dump = spawn(this.LSREGISTER, ['-dump']);
```
Comment (brendankenny) [inline lighthouse-cli/osx-chrome.js]: you'd save a good number of lines of code (and be able to get rid of the `through2` dep) by switching to something like
```javascript
execSync(`${this.LSREGISTER} -dump | grep -i "google chrome canary.app$"`).toString()
```

File: lighthouse-cli/chrome-launcher.js (hunk 9 above)
Comment (boopathi) [inline lighthouse-cli/chrome-launcher.js]: ideally yes... I faced situations where chrome wasn't killed properly, and lighthouse wasn't able to use debugging port and every time I ran lighthouse it kept spawning new chrome instances. so I kept track of all of them.. try it out removing this, if you face this too, lets find why chrome isn't killed properly.

File: lighthouse-cli/osx-chrome.js (hunk 10 above)
Comment (boopathi) [inline lighthouse-cli/osx-chrome.js]: yeah... i'll do that and convert async to sync ...

File: lighthouse-cli/chrome-launcher.js (hunk 5 above)
Comment (brendankenny) [inline lighthouse-cli/chrome-launcher.js]: (and need license on other files below)

File: lighthouse-cli/index.js [hunk 11]
```diff
@@ -139,22 +142,47 @@ if (cli.verbose) {
   flags.logLevel = 'error';
 }
 
+function run(retries) {
```
Comment (brendankenny) [inline lighthouse-cli/index.js]: I like the setup here to default to trying to see if a Chrome we can attach to is already running, but the flow ends up somewhat hard to follow, especially the flow of errors through the conditionals, two `catch` blocks and the recursive `run()` call :)

Can you separate this a bit (into separate functions?) to make the flow clearer? This is only going to get more dire as #640 *also* wants to run this recursively and #639 reworks how errors bubble up to the CLI

File: lighthouse-cli/chrome-launcher.js (hunk 7 above)
Comment (paulirish) [inline lighthouse-cli/chrome-launcher.js]: fwiw i like logging these guys to files.

File: lighthouse-cli/index.js (hunk 3 above)
Comment (boopathi) [inline lighthouse-cli/index.js]: `ask.js` 😃

**Comment by boopathi** [comment]
+ flag `--no-launch-chrome` to disable chrome launch
+ removed through2 and inquirer

**Comment by paulirish** [comment]
I really like the logic of "attempt to use 9222 first", if it connects, we're good. If it fails, launch our own. That smooths out integration in a few places. 👍

File: lighthouse-cli/index.js [hunk 12]
```diff
@@ -62,6 +62,7 @@ const cli = yargs
     'list-all-audits': 'Prints a list of all available audits and exits',
     'list-trace-categories': 'Prints a list of all required trace categories and exits',
     'config-path': 'The path to the config JSON.',
+    'launch-chrome': 'Launch chrome automatically when accessing port 9222 fails',
```
Comment (paulirish) [inline lighthouse-cli/index.js]: Your comments in the PR say `--no-launch-chrome` but the code is just `--launch-chrome`. It looks like you're using http://yargs.js.org/docs/#parsing-tricks-negate-fields to do that? whoa, TIL! very cool.

However, I prefer only really exposing the user to `--no-launch-chrome` or `--no-autolaunch` in `--help` and such. Is that possible?

Comment (boopathi) [inline lighthouse-cli/index.js]: I don't think it's possible. The CLI `--help` will say that it's a boolean and default is true, and to disable it `--no` should be pretty straight forward IMO even if its not obvious.

File: lighthouse-cli/index.js [hunk 13]
```diff
@@ -139,22 +145,73 @@ if (cli.verbose) {
...
+    return launcher
+      .run()
+      .then(() => {
+        console.log('Attempting to run again...');
+        return this.run();
+      })
+      .then(() => launcher.kill())
+      .catch(err => {
+        launcher.kill();
+        this.showRuntimeError(err);
```
Comment (brendankenny) [inline lighthouse-cli/index.js]: since `launchChromeAndRetry` is only called from `run()`, could this be 
```javascript
.catch(err => {
  launcher.kill();
  throw err;
});
```
and `run`'s `catch()` will deal with showing the runtime error?

File: lighthouse-cli/index.js (hunk 12 above)
Comment (boopathi) [inline lighthouse-cli/index.js]: can also use `--launch-chrome=false`

Comment (paulirish) [inline lighthouse-cli/index.js]: I think if we do `"boolean-negation": false` then we can add `no-autolaunch` and `.default('no-autolaunch', false)`

File: lighthouse-cli/index.js [hunk 14]
```diff
@@ -139,22 +145,73 @@ if (cli.verbose) {
...
-    return;
-  })
-  .catch(err => {
-    if (err.code === 'ECONNREFUSED') {
-      console.error('Unable to connect to Chrome. Please run Chrome w/ debugging port 9222 open:');
-      console.error('    npm explore -g lighthouse -- npm run chrome');
-    } else {
-      console.error('Runtime error encountered:', err);
-      console.error(err.stack);
+class Runner {
```
Comment (brendankenny) [inline lighthouse-cli/index.js]: really like the simplification you did here.

My only real feedback would be to drop the class and just have the functions be top level, since that's how this is using them anyway (by immediately instantiating the class and calling `run()` on it)

File: lighthouse-cli/index.js (hunk 12 above)
Comment (paulirish) [inline lighthouse-cli/index.js]: if that's true, we'll have to bump yargs to 4x to use it, i believe.

File: lighthouse-cli/ask.js [hunk 15]
```diff
@@ -0,0 +1,33 @@
+'use strict';
```
Comment (brendankenny) [inline lighthouse-cli/ask.js]: still need licenses for all these files :)

File: lighthouse-cli/index.js (hunk 12 above)
Comment (boopathi) [inline lighthouse-cli/index.js]: Bikeshedding name for this flag - `launch-chrome`, `autolaunch`, `autolaunch-chrome`, `with-debugger`, `with-chrome` ?

File: lighthouse-cli/chrome-launcher.js [hunk 16]
```diff
@@ -0,0 +1,175 @@
...
+      client.unref();
+    }
+  }
+
+  connect() {
+    return new Promise((resolve, reject) => {
+      const client = net.createConnection(9222);
+      client.once('error', err => {
+        this.cleanup(client);
+        reject(err);
```
Comment (paulirish) [inline lighthouse-cli/chrome-launcher.js]: Currently it looks like this if you dont have a browser already open:

image

Can we avoid going so keep into the driver before we have a chrome in place?  Also, is that what this createConnection is about?

File: lighthouse-cli/chrome-finder.js [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+      [/^\/Volumes\//, -1],
+      [/^\/Applications\//, 100],
+      [new RegExp(`^${process.env.HOME}/Applications/`), 50]
+    ]);
+
+    return sort(installations, priorities);
+  },
+
+  linux() {
+    return [process.env.LIGHTHOUSE_CHROMIUM_PATH];
```
Comment (brendankenny) [inline lighthouse-cli/chrome-finder.js]: probably worth copying over [the warning](https://github.com/GoogleChrome/lighthouse/blob/067db41f599503a972cee6e53d9e420fcd106be8/lighthouse-core/scripts/launch-chrome.sh#L46) from `launch-chrome.sh` if this isn't set

File: lighthouse-cli/chrome-finder.js [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+    return installations;
+  }
+};
+
+function sort(installations, priorities) {
+  const defaultPriority = 10;
+  return installations
+    // assign priorities
+    .map(inst => {
+      for (let [regex, priority] of priorities) {
```
Comment (brendankenny) [inline lighthouse-cli/chrome-finder.js]: this isn't going to work on node v4

File: lighthouse-cli/index.js (hunk 12 above)
Comment (paulirish) [inline lighthouse-cli/index.js]: specifying "false" in a cmd line flag is really awkward to me, so I really prefer when the flag itself is only used when you're setting it as truthy.

Basically, apply a flag (with no explicit boolean value) in order to bypass a default.

favorites right now: `skip-autolaunch` `avoid-autolaunch`

File: lighthouse-cli/index.js (hunk 13 above)
Comment (boopathi) [inline lighthouse-cli/index.js]: no. run's catch will not catch this error. 

```js
.catch(err => {
  throw err;
});
```

this error won't be caught. Would need another catch statement.

**Comment by brendankenny** [comment]
linux also needs `--disable-setuid-sandbox`

File: lighthouse-cli/chrome-launcher.js (hunk 16 above)
Comment (boopathi) [inline lighthouse-cli/chrome-launcher.js]: this connect function is used only after chrome is spawned ... Hmmm... maybe I can just check before anything starts - and revert the logic - try to launch chrome first and then do the lighthouse run .... in trying to launch chrome if 9222 is open, skip launching and go directly to lighthouse run.

Comment (paulirish) [inline lighthouse-cli/chrome-launcher.js]: yeahh! i think that's on the 💵

File: lighthouse-cli/index.js (hunk 13 above)
Comment (brendankenny) [inline lighthouse-cli/index.js]: ah, d'oh, you're right, since it's called from inside `run`'s `catch()`

File: lighthouse-cli/ask.js (hunk 15 above)
Comment (boopathi) [inline lighthouse-cli/ask.js]: 👍

File: lighthouse-cli/chrome-launcher.js (hunk 16 above)
Comment (boopathi) [inline lighthouse-cli/chrome-launcher.js]: 👍 pushed those changes

File: lighthouse-cli/chrome-launcher.js (hunk 8 above)
Comment (paulirish) [inline lighthouse-cli/chrome-launcher.js]: I put an extra canary in Downloads so i could try this out.. 

ask() ends up with something like this:
image

(which is pretty cool)

but since `this.head` is set at the top, the 3rd case of the switch statement will always be set and I'm not clear on when ask() is called. (i commented out the 3rd case statement in order to get the above screenshot)

AFAIUI, `head` aka `select-chrome` mean, go with the first canary in your list of multiple, after you've sorted by priority. This basically means, "eh, if you find a few, just pick the one you like the most, lighthouse :)" given that, I think we'd want to tweak both names.

File: lighthouse-cli/chrome-launcher.js (hunk 16 above)
Comment (paulirish) [inline lighthouse-cli/chrome-launcher.js]: Yup this works great now. I love it.

File: lighthouse-cli/chrome-launcher.js [hunk 19]
```diff
@@ -0,0 +1,198 @@
...
+    return flags;
+  }
+
+  prepare() {
+    switch (process.platform) {
+      case 'darwin':
+        this.TMP_PROFILE_DIR = unixTmpDir();
+        break;
+
+      case 'linux':
```
Comment (paulirish) [inline lighthouse-cli/chrome-launcher.js]: wasn't this using a fallthrough before? I kinda prefer that, but no big deal. up to you.

Comment (boopathi) [inline lighthouse-cli/chrome-launcher.js]: yeah... I liked that better too ... Was using two switch statements. Merged to one and forgot to change.

File: lighthouse-cli/chrome-launcher.js (hunk 8 above)
Comment (boopathi) [inline lighthouse-cli/chrome-launcher.js]: yes. Head means exactly that - "I don't care about where you run, take your most preferred one".

`head` and `select-chrome` have an [inverse relation](https://github.com/GoogleChrome/lighthouse/pull/642/files#diff-87dc5351257b763f95f05a89e01fe573R151) ... "If you want to `select-chrome` don't use your top preference `head`" - something like that.

The appropriate things I could think of were - `head`, `maximum`, `top`, `front`, or `pop` - as it was the top of a priority list.

File: lighthouse-cli/ask.js [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+      } else {
+        iface.prompt();
+      }
+    });
+  });
+}
+
+function toInt(n) {
+  try {
+    return parseInt(n, 10);
```
Comment (brendankenny) [inline lighthouse-cli/ask.js]: I don't believe `parseInt` can ever throw. Do you actually want to check for `NaN` here?

File: lighthouse-cli/index.js [hunk 21]
```diff
@@ -29,6 +29,7 @@ const yargs = require('yargs');
 const Printer = require('./printer');
 const lighthouse = require('../lighthouse-core');
 const assetSaver = require('../lighthouse-core/lib/asset-saver.js');
+const Launcher = require('./chrome-launcher');
```
Comment (brendankenny) [inline lighthouse-cli/index.js]: maybe call something like `ChromeLauncher` so it's clear that it's responsible for launching Chrome and not Lighthouse?

File: lighthouse-cli/index.js [hunk 22]
```diff
@@ -91,6 +96,7 @@ Example: --output-path=./lighthouse-results.html`
   .default('mobile', true)
   .default('output', Printer.OUTPUT_MODE.pretty)
   .default('output-path', 'stdout')
+  .default('launch-chrome', true)
```
Comment (brendankenny) [inline lighthouse-cli/index.js]: should be `.default('skip-autolaunch', false)`? (or drop the line completely since we can rely on `undefined` being falsy

File: lighthouse-cli/index.js [hunk 23]
```diff
@@ -140,35 +146,71 @@ if (cli.verbose) {
...
+function launchChromeAndRun() {
+  const launcher = new Launcher({
+    head: !cli.selectChrome,
+  });
+
+  cleanups.push(() => launcher.kill());
+
+  return launcher
+    .connect()
+    .catch(() => {
```
Comment (brendankenny) [inline lighthouse-cli/index.js]: structure here is great :) Nice and readable

File: lighthouse-cli/index.js [hunk 24]
```diff
@@ -140,35 +146,71 @@ if (cli.verbose) {
...
       }
-      process.exit(1);
     });
 }
 
+function showConnectionError() {
+  console.error(
+    'Unable to connect to Chrome. Please run Chrome w/ debugging port 9222 open:'
+  );
+  console.error('    npm explore -g lighthouse -- npm run chrome');
```
Comment (brendankenny) [inline lighthouse-cli/index.js]: do we want to update this since the default way will now just let lighthouse handle launching chrome?

File: package.json [hunk 25]
```diff
@@ -45,6 +45,8 @@
     "handlebars": "^4.0.5",
     "json-stringify-safe": "^5.0.1",
     "jszip": "2.6.0",
+    "mkdirp": "^0.5.1",
+    "rimraf": "^2.5.4",
```
Comment (brendankenny) [inline package.json]: for various dumb institutional reasons, `rimraf` should be set to `"^2.2.8"`

File: lighthouse-cli/chrome-launcher.js [hunk 26]
```diff
@@ -0,0 +1,220 @@
...
+  run() {
+    if (!this.prepared) {
+      this.prepare();
+    }
+
+    return Promise.resolve()
+      .then(() => {
+        const installations = chromeFinder[process.platform]();
+
+        switch (true) {
```
Comment (brendankenny) [inline lighthouse-cli/chrome-launcher.js]: can you switch this switch to using old fashioned `if`/`else` statements? :D

File: lighthouse-cli/chrome-launcher.js (hunk 8 above)
Comment (brendankenny) [inline lighthouse-cli/chrome-launcher.js]: One approach would be to rather than describe what it is, you could describe what it does. Even something as terrible as `useFirstChromeFound` (ignoring prioritization that's actually at work there) would help make its purpose clearer

File: lighthouse-cli/chrome-launcher.js [hunk 27]
```diff
@@ -0,0 +1,220 @@
...
+  cleanup(client) {
+    if (client) {
+      client.removeAllListeners();
+      client.end();
+      client.destroy();
+      client.unref();
+    }
+  }
+
+  connect() {
```
Comment (brendankenny) [inline lighthouse-cli/chrome-launcher.js]: maybe something like `isLaunched()` to be more descriptive?

File: lighthouse-cli/chrome-launcher.js [hunk 28]
```diff
@@ -0,0 +1,220 @@
...
+        reject(err);
+      });
+      client.once('connect', _ => {
+        this.cleanup(client);
+        resolve();
+      });
+    });
+  }
+
+  poll() {
```
Comment (brendankenny) [inline lighthouse-cli/chrome-launcher.js]: `waitForLaunched()` `awaitReady()`? Those are kind of terrible, but just trying to better describe what the method is up to

File: lighthouse-cli/ask.js (hunk 20 above)
Comment (boopathi) [inline lighthouse-cli/ask.js]: Oops. yes.

File: lighthouse-cli/index.js (hunk 22 above)
Comment (boopathi) [inline lighthouse-cli/index.js]: yeah. If it's declared boolean, by default it'll be false rather than undefined.

File: lighthouse-cli/chrome-launcher.js (hunk 28 above)
Comment (boopathi) [inline lighthouse-cli/chrome-launcher.js]: hmm, both names are actually dependent in descriptive form - if the previous one is `isX`, this one should be `waitUntilX` ...

Comment (boopathi) [inline lighthouse-cli/chrome-launcher.js]: So, it could be `isDebuggerReady` and `waitUntilDebuggerReady` ... but it's too long ... So I thought connect and poll would be better because that's what they do ...

File: lighthouse-cli/chrome-launcher.js (hunk 8 above)
Comment (boopathi) [inline lighthouse-cli/chrome-launcher.js]: maybe `useSomeChrome`, `preferSomeChrome`, `autoSelectChrome`? I like `autoSelectChrome` ... descriptive and relatively short ...

File: lighthouse-cli/chrome-launcher.js (hunk 28 above)
Comment (brendankenny) [inline lighthouse-cli/chrome-launcher.js]: agreed that they should be the same `X`. I'm not a fan of really long names, but I'd rather have long names than no description :) For instance, my first question on seeing `poll()` would be "poll for what?" And then there's the fact that as a user of it you don't actually care that it's polling, you only care that you're getting back a promise that's resolved when the driver can be attached to the launched Chrome.

Same thing for `connect()`. To determine if Chrome is ready to be talked to it connects to it, but that should be unobservable from the outside; as a user, you care only that it's helping you determine that Chrome is ready.

Naming things is always hard, so I wouldn't spend too much time coming up with names. I'd personally just go with straightforward but possibly awkwardly long names and we can replace them if we come up with something better in the future.

Comment (boopathi) [inline lighthouse-cli/chrome-launcher.js]: went ahead with `isDebuggerReady` and `waitUntilReady` ... `waitUntilReady` is used after chrome-spawn - so don't have to have long name... `isDebuggerReady` is used outside, so let's use the word Debugger ....

Comment (brendankenny) [inline lighthouse-cli/chrome-launcher.js]: LGTM

**Comment by brendankenny** [comment]
looks like something is going on with node v4 and the smoketests. Timing out instead of passing or failing :(

**Comment by boopathi** [comment]
Yes... Only happens on Node4. trying locally.

**Comment by boopathi** [comment]
Node4 doesn't allow fs.writeFileSync() to pass in a file descriptor and expects only a path.

**Comment by wardpeet** [comment]
@boopathi @paulirish should we create issues for the above flags?
